Getting Currency data from Currency data layers middleware stack

LittleBigDev opened the pull request:
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | **As a Currency Repository, I need a data layer to read/write my currency information**<br>This layer should be able to read from CLDR and from DB.<br>Information from the DB should override the one from CLDR.<br>We should be able to read directly from CLDR in order to install a currency on the DB<br>We need to be able to write on DB too.
| Type?         | new feature
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | [FF-144](http://forge.prestashop.com/browse/FF-144)
| How to test?  | Unit tests suite (`Tests/Unit/Core/Localization/`). <br>⚠️ WARNING : currency data structure has changed (now multilang). Possible side effects on any currency-related feature !! ⚠️

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/8781)
<!-- Reviewable:end -->

LittleBigDev committed:
Added currencies data returned by CLDR reader

LittleBigDev committed:
Currency layers : cache

LittleBigDev committed:
WIP : revamping relations between instances, repos, CLDR, etc.

LittleBigDev committed:
Enhanced Locale repo so it extracts real data from CLDR LocaleData
Also added missing tests, and enhanced some others

LittleBigDev committed:
- Added phpunit coverage whitelist config
- Moved some test classes
- Cleaned some dead code (highlighted by coverage runs)

LittleBigDev committed:
Introducing CurrencyData, used to build Currency instances

LittleBigDev committed:
WIP on currency data source and layers

LittleBigDev committed:
Renamed CLDR layers for better distinction with higher level data layers

LittleBigDev committed:
Updated comments too. Also, fixed little bug.

LittleBigDev committed:
Added missing CLDR data layers' unit tests

LittleBigDev committed:
WIP: higher level CurrencyData layers

LittleBigDev committed:
DataLayer-related classes were a mess. Cleaned up names and folders.

LittleBigDev committed:
Fixed data mapping issue in CurrencyReference data layer

LittleBigDev committed:
DB structure change for currency entity. Now multilang.

LittleBigDev committed:
Fixed currency data retrieval by Currency Reference layer

LittleBigDev committed:
WIP on currency DB layer

LittleBigDev committed:
Fixed tests about CurrencyReference layer

kpodemski commented:
Sooo, thanks to these changes we will be able to change display of the prices from for eg. `11,29` to `11`? this was possible on 1.6 thanks to setting `0` on `PS_PRICE_DISPLAY_PRECISION`

LittleBigDev commented:
Hello @kpodemski, this PR is indeed part of the big plan to reintroduce prices customization.
And yes, precision will be customizable by currency (but not by language).

LittleBigDev committed:
WIP on CurrencyDatabase layer

LittleBigDev committed:
CurrencyDatabase layer : write

LittleBigDev committed:
Better phpDoc + boy scout rule

LittleBigDev committed:
Following Codacy advice about getters naming when they return a boolean

eternoendless commented:
Reviewed 1 of 16 files at r2, 6 of 34 files at r3, 5 of 22 files at r4.
Review status: 11 of 43 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[install-dev/data/db_structure.sql, line 558 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7_-2xtvWEHjEK72y8b:-L7_-2xuacDkCwZOQyfz:bjv3o8v) ([raw file](https://github.com/prestashop/prestashop/blob/a33c3c183931879e17dfb211e232403349b347a1/install-dev/data/db_structure.sql#L558)):*
> ```SQLPL
>   `iso_code` varchar(3) NOT NULL DEFAULT '0',
>   `numeric_iso_code` varchar(3),
>   `precision` INT(2) NOT NULL DEFAULT 2,
> ```

Please use lower case for types, eg `int(2)`

---

*[install-dev/data/db_structure.sql, line 558 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7_-F8n78mcXXkd6ZsT:-L7_-F8n78mcXXkd6ZsU:bc8tqg1) ([raw file](https://github.com/prestashop/prestashop/blob/a33c3c183931879e17dfb211e232403349b347a1/install-dev/data/db_structure.sql#L558)):*
> ```SQLPL
>   `iso_code` varchar(3) NOT NULL DEFAULT '0',
>   `numeric_iso_code` varchar(3),
>   `precision` INT(2) NOT NULL DEFAULT 2,
> ```

I don't think it's a good idea to set an arbitrary low maximum number of decimal digits. Think about bitcoins.

---

*[install-dev/data/db_structure.sql, line 571 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7_-S8gU5qfBG9lsN1W:-L7_-S8gU5qfBG9lsN1X:b-c7tm7y) ([raw file](https://github.com/prestashop/prestashop/blob/a33c3c183931879e17dfb211e232403349b347a1/install-dev/data/db_structure.sql#L571)):*
> ```SQLPL
>   `id_lang` int(10) unsigned NOT NULL,
>   `name` varchar(64) NOT NULL,
>   `symbol` VARCHAR(32) NOT NULL,
> ```

I don't think we should ever set an arbitrary varchar length lower than 255 unless there's a business reason to truncate strings to a fixed length.
Storage wise there's no difference, it takes content length+1 bytes to store regardless of the column size.

---

*[install-dev/data/db_structure.sql, line 571 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7_-W3pe1wDIgTuK4Wj:-L7_-W3pe1wDIgTuK4Wk:bjr4hzv) ([raw file](https://github.com/prestashop/prestashop/blob/a33c3c183931879e17dfb211e232403349b347a1/install-dev/data/db_structure.sql#L571)):*
> ```SQLPL
>   `id_lang` int(10) unsigned NOT NULL,
>   `name` varchar(64) NOT NULL,
>   `symbol` VARCHAR(32) NOT NULL,
> ```

Please use lower case for types

---

*[install-dev/upgrade/sql/1.7.4.0.sql, line 12 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7_4Nsjs8uA8XR43-Y4:-L7_4Nsjs8uA8XR43-Y5:b-febiqn) ([raw file](https://github.com/prestashop/prestashop/blob/a33c3c183931879e17dfb211e232403349b347a1/install-dev/upgrade/sql/1.7.4.0.sql#L12)):*
> ```SQLPL
>   `id_lang` int(10) unsigned NOT NULL,
>   `name` varchar(64) NOT NULL,
>   `symbol` VARCHAR(32) NOT NULL
> ```

Same comment here about arbitrary length

---

*[src/Adapter/Currency/CurrencyDataProvider.php, line 97 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7_6nuv8XYvDvcXVV3Z:-L7_6nuv8XYvDvcXVV3_:b-src90) ([raw file](https://github.com/prestashop/prestashop/blob/a33c3c183931879e17dfb211e232403349b347a1/src/Adapter/Currency/CurrencyDataProvider.php#L97)):*
> ```PHP
>             if (null === $idLang) {
>                 $idLang = Configuration::get('PS_LANG_DEFAULT');
>             }
> ```

This `if` block can be done before `$this->getCurrenyByIsoCode()`

---

*[src/Core/Localization/CLDR/Currency.php, line 170 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7_OiUQUsQnc0j_EXMr:-L7_OiUQUsQnc0j_EXMs:badfaik) ([raw file](https://github.com/prestashop/prestashop/blob/a33c3c183931879e17dfb211e232403349b347a1/src/Core/Localization/CLDR/Currency.php#L170)):*
> ```PHP
>     {
>         if (!in_array($type, [self::SYMBOL_TYPE_NARROW, self::SYMBOL_TYPE_DEFAULT])) {
>             throw new LocalizationException('Unknown symbol type');
> ```

I think it's a good idea to include the requested value in the exception message when throwing this kind of exceptions:

```php
throw new LocalizationException(
    sprintf(
        'Unknown symbol type: "%s"',
        print_r($type, true)
    )
);
```

---

*[src/Core/Localization/CLDR/Locale.php, line 244 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7_Rv6p82dtE5_9UISv:-L7_Rv6p82dtE5_9UISw:b-oo08zf) ([raw file](https://github.com/prestashop/prestashop/blob/a33c3c183931879e17dfb211e232403349b347a1/src/Core/Localization/CLDR/Locale.php#L244)):*
> ```PHP
>         }
>         if (!isset($this->percentPatterns[$numberingSystem])) {
>             throw new LocalizationException('No percent pattern found for numbering system : ' . $numberingSystem);
> ```

Remember that in English there's no space between the end of the word and characters such as `:`, `;`, `!` and `?`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8781)*
<!-- Sent from Reviewable.io -->

LittleBigDev committed:
Minor fixes in SQL, punctuation, code clarification...

LittleBigDev commented:
Review status: 11 of 43 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[install-dev/data/db_structure.sql, line 558 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7_-2xtvWEHjEK72y8b:-L7cgV5Qk2cbSjhM-vZr:bs7wunb) ([raw file](https://github.com/prestashop/prestashop/blob/a33c3c183931879e17dfb211e232403349b347a1/install-dev/data/db_structure.sql#L558)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

Please use lower case for types, eg `int(2)`
</blockquote></details>

Done..

---

*[install-dev/data/db_structure.sql, line 558 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7_-F8n78mcXXkd6ZsT:-L7cgYmJ4j3KnfFi2V2E:b-tvqgj5) ([raw file](https://github.com/prestashop/prestashop/blob/a33c3c183931879e17dfb211e232403349b347a1/install-dev/data/db_structure.sql#L558)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

I don't think it's a good idea to set an arbitrary low maximum number of decimal digits. Think about bitcoins.
</blockquote></details>

I'll set 8 instead.

---

*[install-dev/data/db_structure.sql, line 571 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7_-S8gU5qfBG9lsN1W:-L7cgdL1akzcdEiFJkus:bkxly6x) ([raw file](https://github.com/prestashop/prestashop/blob/a33c3c183931879e17dfb211e232403349b347a1/install-dev/data/db_structure.sql#L571)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

I don't think we should ever set an arbitrary varchar length lower than 255 unless there's a business reason to truncate strings to a fixed length.
Storage wise there's no difference, it takes content length+1 bytes to store regardless of the column size.
</blockquote></details>

Done.
I'll leave the size of 3 for ISO codes though.

---

*[install-dev/data/db_structure.sql, line 571 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7_-W3pe1wDIgTuK4Wj:-L7cgiCB4VAatjJeBgZV:b-896fix) ([raw file](https://github.com/prestashop/prestashop/blob/a33c3c183931879e17dfb211e232403349b347a1/install-dev/data/db_structure.sql#L571)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

Please use lower case for types
</blockquote></details>

Done.

---

*[install-dev/upgrade/sql/1.7.4.0.sql, line 12 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7_4Nsjs8uA8XR43-Y4:-L7cgjynmf6mByda3rAK:b-896fix) ([raw file](https://github.com/prestashop/prestashop/blob/a33c3c183931879e17dfb211e232403349b347a1/install-dev/upgrade/sql/1.7.4.0.sql#L12)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

Same comment here about arbitrary length
</blockquote></details>

Done.

---

*[src/Adapter/Currency/CurrencyDataProvider.php, line 97 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7_6nuv8XYvDvcXVV3Z:-L7cglJVtnxGthV8kxmj:b-896fix) ([raw file](https://github.com/prestashop/prestashop/blob/a33c3c183931879e17dfb211e232403349b347a1/src/Adapter/Currency/CurrencyDataProvider.php#L97)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

This `if` block can be done before `$this->getCurrenyByIsoCode()`
</blockquote></details>

Done.

---

*[src/Core/Localization/CLDR/Currency.php, line 170 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7_OiUQUsQnc0j_EXMr:-L7cgmBrWt9va5Y0VPXJ:b5wy8a3) ([raw file](https://github.com/prestashop/prestashop/blob/a33c3c183931879e17dfb211e232403349b347a1/src/Core/Localization/CLDR/Currency.php#L170)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

I think it's a good idea to include the requested value in the exception message when throwing this kind of exceptions:

```php
throw new LocalizationException(
    sprintf(
        'Unknown symbol type: "%s"',
        print_r($type, true)
    )
);
```
</blockquote></details>

Thanks about the `print_r()` suggestion. 

---

*[src/Core/Localization/CLDR/Locale.php, line 244 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7_Rv6p82dtE5_9UISv:-L7ci8hBZhraFsBNE6_M:bi81fed) ([raw file](https://github.com/prestashop/prestashop/blob/a33c3c183931879e17dfb211e232403349b347a1/src/Core/Localization/CLDR/Locale.php#L244)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

Remember that in English there's no space between the end of the word and characters such as `:`, `;`, `!` and `?`.
</blockquote></details>

Yes it's harder to change your mindset when it's about punctuation :)
I had to fix this mistake in many files of Core/Localization/ 🙈

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8781)*
<!-- Sent from Reviewable.io -->

eternoendless commented:
Reviewed 2 of 29 files at r1, 3 of 34 files at r3, 2 of 22 files at r4, 8 of 14 files at r5.
Review status: 18 of 47 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[install-dev/data/db_structure.sql, line 558 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7_-2xtvWEHjEK72y8b:-L7cv-JcoXTUfqMVSewU:b8rowj1) ([raw file](https://github.com/prestashop/prestashop/blob/a33c3c183931879e17dfb211e232403349b347a1/install-dev/data/db_structure.sql#L558)):*
<details><summary><i>Previously, LittleBigDev (Bastien Bieri) wrote…</i></summary><blockquote>

Done..
</blockquote></details>

Just for the record, I personally prefer upper case for SQL, but since all the other fields are in lower case I'd prefer to be consistent.

---

*[install-dev/data/db_structure.sql, line 558 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7_-F8n78mcXXkd6ZsT:-L7cvPgKUgmB5aKGEhX5:bcvpcdv) ([raw file](https://github.com/prestashop/prestashop/blob/a33c3c183931879e17dfb211e232403349b347a1/install-dev/data/db_structure.sql#L558)):*
<details><summary><i>Previously, LittleBigDev (Bastien Bieri) wrote…</i></summary><blockquote>

I'll set 8 instead.
</blockquote></details>

Oh sorry, I don't know what I was thinking, I mixed up the size of `int` (2 characters so 0-99) with the number of decimal positions. I think PrestaShop currently shows 6 decimals by default in money fields, so I guess that would be a good default value.

---

*[install-dev/data/db_structure.sql, line 571 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7_-S8gU5qfBG9lsN1W:-L7cwE60HhvDm-_X_hUA:bqy67xj) ([raw file](https://github.com/prestashop/prestashop/blob/a33c3c183931879e17dfb211e232403349b347a1/install-dev/data/db_structure.sql#L571)):*
<details><summary><i>Previously, LittleBigDev (Bastien Bieri) wrote…</i></summary><blockquote>

Done.
I'll leave the size of 3 for ISO codes though.
</blockquote></details>

Fair enough

---

*[src/Core/Localization/CLDR/LocaleDataSource.php, line 46 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7d3ev_JsypbGS1qutl:-L7d3ev_JsypbGS1qutm:b-xu31mi) ([raw file](https://github.com/prestashop/prestashop/blob/738a25e9fa49ddfce3ce5a0f4f7350f4b181f635/src/Core/Localization/CLDR/LocaleDataSource.php#L46)):*
> ```PHP
>     /**
>      * LocaleDataSource constructor needs an array of CldrLocaleDataLayer objects.
>      * These layers will be chained and will act as a middleware stack.
> ```

In which order should the layers be sent in the array? Top down or down up?
Since this class only needs a middleware chain, why don't we just provide the topmost element and construct the chain explicitly elsewhere instead of programmatically here?

---

*[src/Core/Localization/CLDR/LocaleDataSource.php, line 84 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7d3M-NdwQCMRnxPWMi:-L7d3M-NdwQCMRnxPWMj:ba8q4ms) ([raw file](https://github.com/prestashop/prestashop/blob/738a25e9fa49ddfce3ce5a0f4f7350f4b181f635/src/Core/Localization/CLDR/LocaleDataSource.php#L84)):*
> ```PHP
> 
>             $layers[$before]->setLowerLayer($thisLayer);
>         }
> ```

Wouldn't it be simpler with a for loop? You wouldn't need to transform the array.

---

*[src/Core/Localization/CLDR/LocaleRepository.php, line 50 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7d5H7xvdoNPHsfWNNQ:-L7d5H7xvdoNPHsfWNNR:b-1gmay) ([raw file](https://github.com/prestashop/prestashop/blob/738a25e9fa49ddfce3ce5a0f4f7350f4b181f635/src/Core/Localization/CLDR/LocaleRepository.php#L50)):*
> ```PHP
> 
>     /**
>      * @param $localeCode
> ```

I guess this is a string?

---

*[src/Core/Localization/CLDR/Reader.php, line 509 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7d5ZAWJ2Wna-FfOxRg:-L7d5ZAWJ2Wna-FfOxRh:bys522y) ([raw file](https://github.com/prestashop/prestashop/blob/738a25e9fa49ddfce3ce5a0f4f7350f4b181f635/src/Core/Localization/CLDR/Reader.php#L509)):*
> ```PHP
>                     /** @var SimplexmlElement $codesMapping */
>                     $codesMapping                 = $codesMapping[0];
>                     $currencyData->numericIsoCode = (int)(string)$codesMapping->attributes()->numeric;
> ```

Are you sure that an iso code can't start with a leading zero?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8781)*
<!-- Sent from Reviewable.io -->

LittleBigDev committed:
Multiple fixes :
- Changed default precision for currency (now 6)
- Simplified LocaleDataSource dependencies (now top layer instead of an array of layers)
- fixed currency data mapping in the CLDR xml reader (numeric ISO code must be a string)

LittleBigDev commented:
Review status: 14 of 47 files reviewed at latest revision, 4 unresolved discussions.

---

*[install-dev/data/db_structure.sql, line 558 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7_-F8n78mcXXkd6ZsT:-L7d6PaRAMujuWCZfpMF:b-896fix) ([raw file](https://github.com/prestashop/prestashop/blob/a33c3c183931879e17dfb211e232403349b347a1/install-dev/data/db_structure.sql#L558)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

Oh sorry, I don't know what I was thinking, I mixed up the size of `int` (2 characters so 0-99) with the number of decimal positions. I think PrestaShop currently shows 6 decimals by default in money fields, so I guess that would be a good default value.
</blockquote></details>

Done.

---

*[src/Core/Localization/CLDR/LocaleDataSource.php, line 46 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7d3ev_JsypbGS1qutl:-L7d7Q4FCx7wL2fF_F1Q:bt6an3e) ([raw file](https://github.com/prestashop/prestashop/blob/738a25e9fa49ddfce3ce5a0f4f7350f4b181f635/src/Core/Localization/CLDR/LocaleDataSource.php#L46)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

In which order should the layers be sent in the array? Top down or down up?
Since this class only needs a middleware chain, why don't we just provide the topmost element and construct the chain explicitly elsewhere instead of programmatically here?
</blockquote></details>

You got a point. 
Here you are.

---

*[src/Core/Localization/CLDR/LocaleDataSource.php, line 84 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7d3M-NdwQCMRnxPWMi:-L7d9HPmaKwd7uqLVaf5:bz94zic) ([raw file](https://github.com/prestashop/prestashop/blob/738a25e9fa49ddfce3ce5a0f4f7350f4b181f635/src/Core/Localization/CLDR/LocaleDataSource.php#L84)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

Wouldn't it be simpler with a for loop? You wouldn't need to transform the array.
</blockquote></details>

Since I changed the `__construct()` dependency, this method is no longer needed. 

---

*[src/Core/Localization/CLDR/LocaleRepository.php, line 50 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7d5H7xvdoNPHsfWNNQ:-L7d6jmyOUKWL0xfpINj:b-tg9ejg) ([raw file](https://github.com/prestashop/prestashop/blob/738a25e9fa49ddfce3ce5a0f4f7350f4b181f635/src/Core/Localization/CLDR/LocaleRepository.php#L50)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

I guess this is a string?
</blockquote></details>

Done.
+ phpDoc enhancement

---

*[src/Core/Localization/CLDR/Reader.php, line 509 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L7d5ZAWJ2Wna-FfOxRg:-L7d9bluTTBo1b6z5W3o:b-tadhat) ([raw file](https://github.com/prestashop/prestashop/blob/738a25e9fa49ddfce3ce5a0f4f7350f4b181f635/src/Core/Localization/CLDR/Reader.php#L509)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

Are you sure that an iso code can't start with a leading zero?
</blockquote></details>

Yes it can. Because it is a string value, not an int. This old piece of code was made for an integer value. Thanks for pointing it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8781)*
<!-- Sent from Reviewable.io -->

LittleBigDev committed:
Fixed currency installation (required "lang" fields made it fail)

LittleBigDev commented:
Hello @eternoendless, car you check the fixes ?

eternoendless commented:
Reviewed 1 of 29 files at r1, 6 of 16 files at r2, 11 of 34 files at r3, 13 of 22 files at r4, 5 of 14 files at r5, 6 of 6 files at r6, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[classes/Currency.php, line 145 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L823hkTDmcAOzVkVIW6:-L823hkTDmcAOzVkVIW7:bpoi6te) ([raw file](https://github.com/prestashop/prestashop/blob/a33c3c183931879e17dfb211e232403349b347a1/classes/Currency.php#L145)):*
> ```PHP
>             /* Lang fields */
>             'name'   => array('type' => self::TYPE_STRING, 'lang' => true, 'validate' => 'isGenericName', 'required' => true, 'size' => 64),
>             'symbol' => array('type' => self::TYPE_STRING, 'lang' => true, 'required' => true, 'size' => 64),
> ```

Why are these no longer required?

---

*[src/Core/Localization/CLDR/DataLayer/CurrencyCache.php, line 96 at r7](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L823yCgZiCJhVL0Mbeo:-L823yCgZiCJhVL0Mbep:b-dxkll5) ([raw file](https://github.com/prestashop/prestashop/blob/bc79983e265642bc94ed4dcd4f9b8845cacd679c/src/Core/Localization/CLDR/DataLayer/CurrencyCache.php#L96)):*
> ```PHP
>         if (!($data instanceof CurrencyData)) {
>             throw new LocalizationException(
>                 '$data must be an instance of PrestaShop\PrestaShop\Core\Localization\CLDR\CurrencyData'
> ```

You should use `CurrencyData::class` instead of writing the FQCN

---

*[src/Core/Localization/Currency/CurrencyDataSource.php, line 80 at r7](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L82TFUQSWVpHpVChfcE:-L82TFUQSWVpHpVChfcF:b-sdscgn) ([raw file](https://github.com/prestashop/prestashop/blob/bc79983e265642bc94ed4dcd4f9b8845cacd679c/src/Core/Localization/Currency/CurrencyDataSource.php#L80)):*
> ```PHP
>         while ($thisLayer = array_pop($layers)) {
>             $before = count($layers) - 1;
>             if ($before < 0) {
> ```

You can merge the previous two lines in one: `if (0 === count($layers))`

---

*[src/Core/Localization/Currency/CurrencyDataSource.php, line 87 at r7](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L82Tb8kvX8FTQUp0zGc:-L82Tb8kvX8FTQUp0zGd:bfs643a) ([raw file](https://github.com/prestashop/prestashop/blob/bc79983e265642bc94ed4dcd4f9b8845cacd679c/src/Core/Localization/Currency/CurrencyDataSource.php#L87)):*
> ```PHP
>         }
> 
>         return null;
> ```

I think it should throw an exception if `$layers` is empty.

---

*[src/Core/Localization/Currency/DataLayer/CurrencyCache.php, line 96 at r7](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L82U-9VFbHp85gUmvl_:-L82U-9VFbHp85gUmvla:bnepzsr) ([raw file](https://github.com/prestashop/prestashop/blob/bc79983e265642bc94ed4dcd4f9b8845cacd679c/src/Core/Localization/Currency/DataLayer/CurrencyCache.php#L96)):*
> ```PHP
>         if (!($data instanceof CurrencyData)) {
>             throw new LocalizationException(
>                 '$data must be an instance of PrestaShop\PrestaShop\Core\Localization\Currency\CurrencyData'
> ```

Again, you should use CurrencyData::class instead of writing the FQCN

---

*[src/Core/Localization/Currency/DataLayer/CurrencyReference.php, line 55 at r7](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L82UsQYuXIAO4VhH0Is:-L82UsQYuXIAO4VhH0It:bubyz32) ([raw file](https://github.com/prestashop/prestashop/blob/bc79983e265642bc94ed4dcd4f9b8845cacd679c/src/Core/Localization/Currency/DataLayer/CurrencyReference.php#L55)):*
> ```PHP
>     protected $localeCode;
> 
>     public function __construct(CldrLocaleRepository $cldrLocaleRepository, $localeCode)
> ```

Missing Phpdoc (mainly for `$localeCode`)

---

*[src/Core/Localization/Locale/Repository.php, line 127 at r7](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L82WUasOtcOt7o6pOSu:-L82WUasOtcOt7o6pOSv:btx6gud) ([raw file](https://github.com/prestashop/prestashop/blob/bc79983e265642bc94ed4dcd4f9b8845cacd679c/src/Core/Localization/Locale/Repository.php#L127)):*
> ```PHP
>         $currencyDisplayType = PriceSpecification::CURRENCY_DISPLAY_SYMBOL,
>         $groupingUsed = true,
>         $maxFractionDigits = self::MAX_FRACTION_DIGITS
> ```

Don't these two properties depend on the locale itself?

---

*[src/Core/Localization/Locale/Repository.php, line 276 at r7](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L82X1KCRkZOgVuuQoBf:-L82X1KCRkZOgVuuQoBg:b9qj8k2) ([raw file](https://github.com/prestashop/prestashop/blob/bc79983e265642bc94ed4dcd4f9b8845cacd679c/src/Core/Localization/Locale/Repository.php#L276)):*
> ```PHP
>             $currency->getSymbol($localeCode),
>             $currency->getIsoCode()
>         );
> ```

Looks like this is a complex process that relies on good knowledge of CLDR. How about extracting this process in a factory? It will allow us to get rid of all the methods below that don't really belong to a repository.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8781)*
<!-- Sent from Reviewable.io -->

LittleBigDev committed:
Extracted number spec building from the repository (created factory)
Also enhanced phpDoc and some exceptions messages.

LittleBigDev commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[classes/Currency.php, line 145 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L823hkTDmcAOzVkVIW6:-L82aDzMrdpHccecclV1:bz2tyuy) ([raw file](https://github.com/prestashop/prestashop/blob/a33c3c183931879e17dfb211e232403349b347a1/classes/Currency.php#L145)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

Why are these no longer required?
</blockquote></details>

Because of multilang fields + `validateRules()` bug in the currency installation form. It was the fastest way to fix this, as these value are hard set in the Currency entity constructor anyway. This is obviously temporary until we work on the new currency install / update methods.

---

*[src/Core/Localization/CLDR/DataLayer/CurrencyCache.php, line 96 at r7](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L823yCgZiCJhVL0Mbeo:-L82asG1FALBT1UW1bqo:b-896fix) ([raw file](https://github.com/prestashop/prestashop/blob/bc79983e265642bc94ed4dcd4f9b8845cacd679c/src/Core/Localization/CLDR/DataLayer/CurrencyCache.php#L96)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

You should use `CurrencyData::class` instead of writing the FQCN
</blockquote></details>

Done.

---

*[src/Core/Localization/Currency/CurrencyDataSource.php, line 80 at r7](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L82TFUQSWVpHpVChfcE:-L82b_lAI-eTlhQxw_b6:b-j307di) ([raw file](https://github.com/prestashop/prestashop/blob/bc79983e265642bc94ed4dcd4f9b8845cacd679c/src/Core/Localization/Currency/CurrencyDataSource.php#L80)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

You can merge the previous two lines in one: `if (0 === count($layers))`
</blockquote></details>

Actually this code was forgotten. It should not exist. Please see my fix.

---

*[src/Core/Localization/Currency/CurrencyDataSource.php, line 87 at r7](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L82Tb8kvX8FTQUp0zGc:-L82oMGVIpczxTIgQwKS:b-otilcs) ([raw file](https://github.com/prestashop/prestashop/blob/bc79983e265642bc94ed4dcd4f9b8845cacd679c/src/Core/Localization/Currency/CurrencyDataSource.php#L87)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

I think it should throw an exception if `$layers` is empty.
</blockquote></details>

See comment above.

---

*[src/Core/Localization/Currency/DataLayer/CurrencyCache.php, line 96 at r7](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L82U-9VFbHp85gUmvl_:-L82c45G15Kbwrt7iCa5:b-896fix) ([raw file](https://github.com/prestashop/prestashop/blob/bc79983e265642bc94ed4dcd4f9b8845cacd679c/src/Core/Localization/Currency/DataLayer/CurrencyCache.php#L96)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

Again, you should use CurrencyData::class instead of writing the FQCN
</blockquote></details>

Done.

---

*[src/Core/Localization/Currency/DataLayer/CurrencyReference.php, line 55 at r7](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L82UsQYuXIAO4VhH0Is:-L82cPcLEMJ4qI7tmtlq:b-896fix) ([raw file](https://github.com/prestashop/prestashop/blob/bc79983e265642bc94ed4dcd4f9b8845cacd679c/src/Core/Localization/Currency/DataLayer/CurrencyReference.php#L55)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

Missing Phpdoc (mainly for `$localeCode`)
</blockquote></details>

Done.

---

*[src/Core/Localization/Locale/Repository.php, line 127 at r7](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L82WUasOtcOt7o6pOSu:-L82ccbmVmcp3exKL6M8:b-yqxikc) ([raw file](https://github.com/prestashop/prestashop/blob/bc79983e265642bc94ed4dcd4f9b8845cacd679c/src/Core/Localization/Locale/Repository.php#L127)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

Don't these two properties depend on the locale itself?
</blockquote></details>

CLDR locale definition has no strict rule about either rounding mode, used numbering system, currency display type , digits grouping or rounding precision.
When we DECIDE to use it or not (grouping, currency symbol type, etc.) or when we decide the value (rounding mode, precision), then CLDR has indications on how to do it well. But there is no obligation. That's why these options must be defined somewhere else. I set default values, and then all of them should be configurable in admin.

---

*[src/Core/Localization/Locale/Repository.php, line 276 at r7](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L82X1KCRkZOgVuuQoBf:-L82oJmPSzj-1MAHCM2t:b-896fix) ([raw file](https://github.com/prestashop/prestashop/blob/bc79983e265642bc94ed4dcd4f9b8845cacd679c/src/Core/Localization/Locale/Repository.php#L276)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

Looks like this is a complex process that relies on good knowledge of CLDR. How about extracting this process in a factory? It will allow us to get rid of all the methods below that don't really belong to a repository.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8781)*
<!-- Sent from Reviewable.io -->

LittleBigDev commented:
@eternoendless here we are :)

eternoendless commented:
Reviewed 4 of 34 files at r3, 8 of 8 files at r8.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[classes/Currency.php, line 145 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L823hkTDmcAOzVkVIW6:-L82pjyl8SHOarRDOPKL:b-w9sxr8) ([raw file](https://github.com/prestashop/prestashop/blob/a33c3c183931879e17dfb211e232403349b347a1/classes/Currency.php#L145)):*
<details><summary><i>Previously, LittleBigDev (Bastien Bieri) wrote…</i></summary><blockquote>

Because of multilang fields + `validateRules()` bug in the currency installation form. It was the fastest way to fix this, as these value are hard set in the Currency entity constructor anyway. This is obviously temporary until we work on the new currency install / update methods.
</blockquote></details>

Ok then

---

*[src/Core/Localization/Currency/CurrencyDataSource.php, line 80 at r7](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L82TFUQSWVpHpVChfcE:-L82pudArl05fKM9FoYl:b-rx4utx) ([raw file](https://github.com/prestashop/prestashop/blob/bc79983e265642bc94ed4dcd4f9b8845cacd679c/src/Core/Localization/Currency/CurrencyDataSource.php#L80)):*
<details><summary><i>Previously, LittleBigDev (Bastien Bieri) wrote…</i></summary><blockquote>

Actually this code was forgotten. It should not exist. Please see my fix.
</blockquote></details>

It still exists :)

---

*[src/Core/Localization/Locale/Repository.php, line 127 at r7](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-L82WUasOtcOt7o6pOSu:-L82qACfkioXPwL9FFWM:b-60carf) ([raw file](https://github.com/prestashop/prestashop/blob/bc79983e265642bc94ed4dcd4f9b8845cacd679c/src/Core/Localization/Locale/Repository.php#L127)):*
<details><summary><i>Previously, LittleBigDev (Bastien Bieri) wrote…</i></summary><blockquote>

CLDR locale definition has no strict rule about either rounding mode, used numbering system, currency display type , digits grouping or rounding precision.
When we DECIDE to use it or not (grouping, currency symbol type, etc.) or when we decide the value (rounding mode, precision), then CLDR has indications on how to do it well. But there is no obligation. That's why these options must be defined somewhere else. I set default values, and then all of them should be configurable in admin.
</blockquote></details>

Ok then!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8781)*
<!-- Sent from Reviewable.io -->

LittleBigDev committed:
Deleted dead code

LittleBigDev commented:
Review status: 48 of 49 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/Core/Localization/Currency/CurrencyDataSource.php, line 80 at r7](https://beta.reviewable.io/reviews/prestashop/prestashop/8781#-L82TFUQSWVpHpVChfcE:-L82r46A7FQtwcED6Q6w:b-1u0x3a) ([raw file](https://github.com/prestashop/prestashop/blob/bc79983e265642bc94ed4dcd4f9b8845cacd679c/src/Core/Localization/Currency/CurrencyDataSource.php#L80)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

It still exists :)
</blockquote></details>

Done.
My bad...

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/prestashop/prestashop/8781)*
<!-- Sent from Reviewable.io -->

eternoendless commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 34 files at r3, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8781#-:-L82t-TpyKqg_laSI_Rg:bnfp4nl)*
<!-- Sent from Reviewable.io -->
